cpp/src/arrow/python/arrow_to_python.cc

```diff
@@ -187,6 +195,16 @@ Status DeserializeTuple(PyObject* context, std::shared_ptr<Array> array,
   DESERIALIZE_SEQUENCE(PyTuple_New, PyTuple_SET_ITEM);
 }
 
+#define SET_CREATE_FN(SIZE) PySet_New(nullptr)
+#define SET_SET_ITEM_FN(SET, INDES, VALUE) PySet_Add(SET, VALUE)
```

Comment:
`PySet_Add` can fail where the other SET_ITEM options cannot. I think it's OK to in-line the contents of the macro here and add appropriate error checks